From a3852227ea01cfc3db52f9a791799eb1444a2bb9 Mon Sep 17 00:00:00 2001
From: Mike Snitzer <snitzer@redhat.com>
Date: Fri, 28 Nov 2014 16:18:35 -0500
Subject: [md] dm-cache: discard block size must be a multiple of cache block size

Message-id: <1417191518-16689-7-git-send-email-snitzer@redhat.com>
Patchwork-id: 101008
O-Subject: [RHEL7.1 PATCH 6/9] dm cache: discard block size must be a multiple of cache block size
Bugzilla: 1165050
RH-Acked-by: Joe Thornber <thornber@redhat.com>
RH-Acked-by: Heinz Mauelshagen <heinzm@redhat.com>

BZ: 1165050

Upstream linux-dm.git commit aa67434ee40c9050b2d291045e124b10607ce0eb
Author: Joe Thornber <ejt@redhat.com>
Date:   Wed Nov 26 16:07:50 2014 +0000

    dm cache: discard block size must be a multiple of cache block size

    Otherwise the cache blocks may span two discard blocks, which we don't
    handle when doing the discard lookup.

    Signed-off-by: Joe Thornber <ejt@redhat.com>
    Signed-off-by: Mike Snitzer <snitzer@redhat.com>

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/md/dm-cache-target.c b/drivers/md/dm-cache-target.c
index 894648d..856d0c0 100644
--- a/drivers/md/dm-cache-target.c
+++ b/drivers/md/dm-cache-target.c
@@ -2269,9 +2269,8 @@ static int create_cache_policy(struct cache *cache, struct cache_args *ca,
 }
 
 /*
- * We want the discard block size to be a power of two, at least the size
- * of the cache block size, and have no more than 2^14 discard blocks
- * across the origin.
+ * We want the discard block size to be at least the size of the cache
+ * block size and have no more than 2^14 discard blocks across the origin.
  */
 #define MAX_DISCARD_BLOCKS (1 << 14)
 
@@ -2286,9 +2285,7 @@ static bool too_many_discard_blocks(sector_t discard_block_size,
 static sector_t calculate_discard_block_size(sector_t cache_block_size,
           sector_t origin_size)
 {
- sector_t discard_block_size;
-
- discard_block_size = roundup_pow_of_two(cache_block_size);
+ sector_t discard_block_size = cache_block_size;
 
  if (origin_size)
   while (too_many_discard_blocks(discard_block_size, origin_size))
-- 
1.7.1